specifyweb/frontend/js_src/lib/components/Header/Notifications.tsx

```diff
@@ -85,34 +99,47 @@ export function Notifications({
             }
           )
         )
-        .then(({ data: notifications }) => {
+        .then(({ data: newNotifications }) => {
           if (destructorCalled) return undefined;
-          setNotifications(
-            notifications
-              .map(
-                // eslint-disable-next-line @typescript-eslint/naming-convention
-                ({ message_id, read, timestamp, type, ...rest }) => ({
-                  messageId: message_id,
-                  read,
-                  timestamp,
-                  type,
-                  payload: rest as IR<LocalizedString>,
-                })
-              )
-              // Make most recent notification first
-              .sort(
-                sortFunction(
-                  ({ timestamp }) => new Date(timestamp).getTime(),
-                  true
+
+          setNotifications((existingNotifications) => {
+            const mappedNewNotifications = newNotifications.map(
+              ({ message_id, read, timestamp, type, ...rest }) => ({
+                messageId: message_id,
+                read,
+                timestamp,
+                type,
+                payload: rest as IR<LocalizedString>,
+              })
+            );
+
+            const filteredNewNotifications = mappedNewNotifications.filter(
+              (newNotification) =>
+                !existingNotifications?.some(
+                  (existingNotification) =>
+                    existingNotification.messageId === newNotification.messageId
                 )
+            );
+
+            return [
+              ...(existingNotifications ?? []),
+              ...filteredNewNotifications,
+            ].sort(
+              sortFunction(
+                ({ timestamp }) => new Date(timestamp).getTime(),
+                true
               )
-          );
+            );
+          });
+          lastFetchedTimestamp = startFetchTimestamp;
           // Stop updating if tab is hidden
-          timeout =
+          return (timeout =
```

Comment:
this is not very readable - try to avoid assigning and returning on the same line